Introduce UserException

mlocati opened the pull request:
We sometimes throw an exception when there's something wrong, and sometimes we would like to show the exception message to the site visitors (for instance: *"Please enter an integer greater than 0"*).

The ecosystem too can throw an exception when something wrong happens, but we don't want these errors to be shown to site visitors (if `concrete.debug.display_errors` is set to `false`).

So, we need a way to say *"Hey, please show this error to visitors, regardless the `concrete.debug.display_errors` option"*, but this currently may be overcomplicated.

So, what about adding a new `UserException` class, representing errors that should be shown to visitors?

hissy commented:
Nice addition. 👍

KorvinSzanto commented:
We created a branch for this in #4487, no?
On Tue, Jun 6, 2017 at 7:08 PM Takuro Hishikawa <notifications@github.com>
wrote:

> Nice addition. 👍
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/concrete5/concrete5/pull/5522#issuecomment-306665463>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA9fOx59WUakWMbZi-ZFkdVMpXuyevLNks5sBgYegaJpZM4NwiIM>
> .
>

KorvinSzanto commented:
Err #4816
On Tue, Jun 6, 2017 at 7:50 PM Korvin Szanto <korvinszanto@gmail.com> wrote:

> We created a branch for this in #4487, no?
> On Tue, Jun 6, 2017 at 7:08 PM Takuro Hishikawa <notifications@github.com>
> wrote:
>
>> Nice addition. 👍
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/concrete5/concrete5/pull/5522#issuecomment-306665463>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AA9fOx59WUakWMbZi-ZFkdVMpXuyevLNks5sBgYegaJpZM4NwiIM>
>> .
>>
>

mlocati commented:
@KorvinSzanto This is a different stuff. This is only meant to always show the exception message to the user, regardless the values of the `concrete.debug.display_errors`/`concrete.debug.detail` configuration options.
I renamed the exception from `UserException` to `UserMessageException` to better represent it, and added a couple of commits to showcase its usage.
